src/main/java/persistence/session/SessionImpl.java

```diff
@@ -0,0 +1,173 @@
+package persistence.session;
+
+import persistence.action.ActionQueue;
+import persistence.entity.CollectionPersister;
+import persistence.entity.EntityEntry;
+import persistence.entity.EntityKey;
+import persistence.entity.EntityPersister;
+import persistence.entity.PersistenceContext;
+import persistence.event.EventSource;
+import persistence.event.SessionService;
+import persistence.event.delete.DeleteEvent;
+import persistence.event.delete.DeleteEventListener;
+import persistence.event.flush.FlushEvent;
+import persistence.event.flush.FlushEventListener;
+import persistence.event.load.LoadEvent;
+import persistence.event.load.LoadEventListener;
+import persistence.event.merge.MergeEvent;
+import persistence.event.merge.MergeEventListener;
+import persistence.event.persist.PersistEvent;
+import persistence.event.persist.PersistEventListener;
+import persistence.meta.Metamodel;
+import persistence.sql.definition.TableAssociationDefinition;
+
+import java.io.Serializable;
+import java.util.function.Supplier;
+
+public class SessionImpl implements EventSource {
+    private final PersistenceContext persistenceContext;
+    private final Metamodel metamodel;
+    private final SessionService sessionService;
+    private final ActionQueue actionQueue;
+
+    public SessionImpl(PersistenceContext persistenceContext,
+                       Metamodel metamodel,
+                       SessionService sessionService,
+                       ActionQueue actionQueue) {
+
+        this.persistenceContext = persistenceContext;
+        this.metamodel = metamodel;
+        this.sessionService = sessionService;
+        this.actionQueue = actionQueue;
+    }
+
+    @Override
+    public <T> T find(Class<T> clazz, Object id) {
+        final EntityKey entityKey = new EntityKey((Long) id, clazz);
+        final EntityEntry entityEntry = getEntityEntryOrDefault(entityKey, () -> EntityEntry.loading((Serializable) id));
+
+        if (entityEntry.isManaged()) {
+            return clazz.cast(persistenceContext.getEntity(entityKey));
+        }
+
+        check(entityEntry.isNotReadable(), "Entity is not managed: " + clazz.getSimpleName());
+
+        final T entity = metamodel.findEntityLoader(clazz).loadEntity(clazz, entityKey);
+        sessionService.LOAD.fireEventOnEachListener(
+                LoadEvent.create(this, (Serializable) id, entity, entityEntry),
+                LoadEventListener::onLoad
+        );
```

Comment:
```suggestion
        final T entity = metamodel.findEntityLoader(clazz).loadEntity(clazz, entityKey);
        sessionService.LOAD.fireEventOnEachListener(
                LoadEvent.create(this, (Serializable) id, entity, entityEntry),
                LoadEventListener::onLoad
        );
```

코드를 보다가 한가지 궁금한점이 생겨서 질문드려요~! 😄  
이벤트 패턴을 추가하면서 어떤 문제를 해소하려고 하셨는지, 그리고 그 해결 방식이 기존 구조에 어떤 이점을 가져왔다고 생각하시는지 궁금합니다~ 😄

Reply:
넵 ㅎㅎㅎ 제가 했던 생각을 말씀드리면,

이벤트 패턴의 장점은 멀티캐스팅이라고 생각을 했었고, 저희가 사용하는 EntityManager의 대표 기능들에 대해서 여러 부수 처리들을 한번의 발송으로 그룹화하여 처리할 수 있다는 점을 염두해 두었습니다.

기존 구조에서는 (물론 현재는 필요한 기능들만 구현해서 복잡하지는 않았었습니다만) 로직이 추가될때마다 SessionImpl 클래스가 복잡해지기 때문에 추가적인 서비스 역할을 할 수 있는 객체가 필요하였고, 이벤트 형태로 만들어서 실질적으로 리스너들이 어떠한 처리를 하는 지에 대해서 SessionImpl에 숨길 수 있는 장점이 있다고 생각합니다.

또한 이벤트 형태이기 때문에 각 리스너들은 자기 관심사만 처리하면 되어서 단일 책임 원칙을 지키기도 쉬워지는 구조라고 생각합니다.

다만 Hibernate의 코드를 보니 SessionImpl에서는 트랜잭션 관리 정도의 역할만 하고 리스너에게 모두 위임하는 것으로 보여서 기본적으로 요청이 타당한가 의 검증정도는 SessionImpl에 남겨두었습니다.

요약하자면,
SessionImpl에서 부수 로직들에 대해 감추고, 부수 로직들을 처리하는 EventHandler들이 group화되어 관리되기 때문에 특정 Event가 발생하였을 때 각 Handler들이 자기 관심사만 처리할 수 있는 구조가 만들어졌다고 생각합니다. :)

Reply:
좋은 답변 감사합니다! 저도 많은 점을 배우게 되네요 😄 

> SessionImpl에서 부수 로직들에 대해 감추고, 부수 로직들을 처리하는 EventHandler들이 group화되어 관리되기 때문에 특정 Event가 발생하였을 때 각 Handler들이 자기 관심사만 처리

그렇다면, `final T entity = metamodel.findEntityLoader(clazz).loadEntity(clazz, entityKey);` 해당 코드 또한 `SessionImpl` 관심사에서 좀 더 때어내면 어떨까 생각하는데 어떻게 생각하시는지 궁금하네요! 🤔

Reply:
저는 이벤트 발행을 중앙에서 관리하면서 트랜잭션, 비동기 등 처리해야 하는 작업이나 추가 로직은 별도로 분리하여 연계작업을 효과적으로 처리할 수 있는 구조로 바라보고 있어요 ㅎ 🤔

Reply:
> 그렇다면, final T entity = metamodel.findEntityLoader(clazz).loadEntity(clazz, entityKey); 해당 코드 또한 SessionImpl 관심사에서 좀 더 때어내면 어떨까 생각하는데 어떻게 생각하시는지 궁금하네요! 🤔

넵 말씀해주신 부분 실제로 고민 했던 부분입니다. 

분리하는 것이 더 좋은 방향이라고 생각하긴 하는데, 그렇게 되면 EventListener들이 수행이 된 후에 결과값을 받아오기가 애매해지더라구요 ㅎㅎㅎ... 

사실 Event의 반환값을 받아서 사용한다는 것이 조금 어색해서.. + Event가 그룹(Collection)으로 등록되어 순환되며 돌아가는 구조이기 때문에 한번의 메서드 실행으로 조회된 Entity를 반환받는 것이 조금 어렵다고 생각했었어서 반영을 못했었는데요.

실제 하이버네이트의 코드를 조금 살펴보니 아래와 같이 구현이 되어 있긴 하네요.

```
			final EventSource session = event.getSession();
			final EntityKey keyToLoad = session.generateEntityKey( event.getEntityId(), persister );
			if ( loadType.isNakedEntityReturned() ) {
				//do not return a proxy!
				//(this option indicates we are initializing a proxy)
				event.setResult( load( event, persister, keyToLoad, loadType ) );
			}
			else {
				//return a proxy if appropriate
				Object result = event.getLockMode() == LockMode.NONE
						? proxyOrLoad( event, persister, keyToLoad, loadType )
						: lockAndLoad( event, persister, keyToLoad, loadType );
				event.setResult( result );
			}
```

그래서 아까 제가 말씀드린 고민과 로드된 엔티티를 Session에서 어떻게 받아와서 사용할까에 대한 답으로 우선 Event에 결과를 바인딩 하는 구조로 한번 반영해보았습니다. 

set...을 사용하긴 하는데, 어쩔수 없는 부분인 것 같은 느낌이 들긴 하네요.
이부분 반영하였는데 의견 부탁드립니다 :)

Reply:
고민해보셨군요 👍 

저도 사실 하이버네이트 구조가 좋은구조라고도 생각하지 않고 완벽하다고도 생각하지 않아요. ㅎ
다만, 책임과 역할에 대해서 한번더 이야기 나누어 보고 싶었습니다! 😄 
> 어쩌면 Event라는 특수한 도메인 용어를 사용해서 로직의 본질을 힘들게 만드는게 아닐까 생각이 들기도해요

Reply:
고민해보셨군요 👍 

저도 사실 하이버네이트 구조가 좋은구조라고도 생각하지 않고 완벽하다고도 생각하지 않아요. ㅎ
다만, 책임과 역할에 대해서 한번더 이야기 나누어 보고 싶었습니다! 😄 

> 어쩌면 Event라는 용어때문에 많이 헷갈리는것 같긴해요 저도 이구간은 너무 이해하기 힘들더라구요 ㅎ 🤔